### PR TITLE
Implement Network section

### PR DESCRIPTION
**Pull request by jvfdunkley**
### :pushpin: References

-   **Issue:** Closes [Implement network section as button on top right corner](https://app.clickup.com/t/g55m9f)

### :memo: Implementation
- Copied over the structure and CSS from the original network tab
- Made Tools button a dropdown and Network a sub-menu
- Broke down each part of the form into its own component for (hopefully) better readability

### :art: Screenshots
Screen Shot 2021-03-29 at 11 39 08 AM

### :bookmark_tabs: Others

-   [ ] Any change in the [API repo](https://github.com/EyeSeeTea/d2-api)? If so, what branch/PR?
-   [ ] Any change in the [GUI library](https://github.com/EyeSeeTea/d2-ui-components)? If so, what branch/PR?

### CODE REVIEW

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/IncludeNeighborsCheckbox.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+const IncludeNeighborsCheckbox: React.FC<IncludeNeighborsCheckboxProps> = React.memo(props => {
+    const { checkedValue, onCheckboxChange } = props;
+    return (
+        <div className="include-neighbors-checkbox">
+            <input
+                type="checkbox"
+                className="include-neighbors-checkbox"
+                checked={checkedValue}
+                onChange={() => onCheckboxChange()}
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/IncludeNeighborsCheckbox.tsx:18]: Better: `onChange={onCheckboxChange}`, which also prevents a new arrow function to be created and the prop remains immutable (important for pure components performance)

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/SpeciesSelect.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+export interface SpeciesSelectProps {
+    value: string;
+    onSpecieChange: (e: string) => void;
+}
+
+type NetworkIndex = typeof indexValues[number];
+
+const indexTranslations: Record<NetworkIndex, string> = {
+    arabidopsisThaliana: i18n.t("Arabidopsis thaliana"),
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/SpeciesSelect.tsx]: Scientific names, we won't need translations.

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkExample.tsx [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+const NetworkExample: React.FC<NetworkExampleProps> = React.memo(props => {
+    const { onExampleClick } = props;
+    return (
+        <span id="network-example" onClick={() => onExampleClick("P01111 \nP01112 \nP01116")}>
+            <small>Example</small>
+        </span>
+    );
+});
+
+export default NetworkExample;
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkExample.tsx:17]: As of late, we export the component without `default`. No need to change it.

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkExample.tsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+interface NetworkExampleProps {
+    onExampleClick: (e: string) => void;
+}
+
+const NetworkExample: React.FC<NetworkExampleProps> = React.memo(props => {
+    const { onExampleClick } = props;
+    return (
+        <span id="network-example" onClick={() => onExampleClick("P01111 \nP01112 \nP01116")}>
+            <small>Example</small>
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkExample.tsx:12]: i18n.t

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/Label.tsx [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import React from "react";
+import i18n from "../../utils/i18n";
+import "./Network.css";
+
+interface LabelProps {
+    forText: string;
+    labelText: string;
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/Label.tsx:7]: [minor] Label is implicitly a text, and in TS we have types, so usually we'd write just `label: string`.

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkExample.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import "./Network.css";
+
+interface NetworkExampleProps {
+    onExampleClick: (e: string) => void;
+}
+
+const NetworkExample: React.FC<NetworkExampleProps> = React.memo(props => {
+    const { onExampleClick } = props;
+    return (
+        <span id="network-example" onClick={() => onExampleClick("P01111 \nP01112 \nP01116")}>
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkExample.tsx:11]: I guess we can remove the trailing space

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    includeNeighboursWithStructuralData: boolean;
+}
+
+function getInitialNetworkForm(): NetworkForm {
+    return {
+        species: "homoSapiens",
+        uniProtAccession: "",
+        includeNeighboursWithStructuralData: false,
+    };
+}
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx:23]: When no arguments are required to build the initial value, we can simply: `const initialNetworkForm: NetworkFrom = {...}`

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+const NetworkForm = React.memo(() => {
+    const annotationFileRef = useRef<DropzoneRef>(null);
+    const [error, setError] = useState<string>();
+    const [networkForm, setNetworkForm] = useState<NetworkForm>(() => getInitialNetworkForm());
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx:28]: `() => fn()` -> `fn` (Caveat: unless `fn` is in fact a class-method that carries a bound context for `this`!)

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+    const addNetwork = useCallback(() => {
+        setError("");
+        if (networkForm.uniProtAccession === "") {
+            setError("Error: Please write down the UniProt accession");
+        }
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx:34]: Add `TODO` somewhere in the code so we know at a glance that this is a temporal callback.

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    }, [networkForm]);
+
+    return (
+        <form className="network-form">
+            <Label forText="species" labelText="Select species" />
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx:40]: i18n.t

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            <SpeciesSelect
+                value={networkForm.species}
+                onSpecieChange={e =>
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx]: `e` is too short, event? but's not the event, it's the species. Maybe `species` or `newSpecies`.

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+                    setNetworkForm({
+                        ...networkForm,
+                        species: e,
+                    })
+                }
+            />
+            <Label forText="uniProtAccession" labelText="Enter a list of UniProt accession" />
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx:50]: i18n.t

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+                checkedValue={networkForm.includeNeighboursWithStructuralData}
+                onCheckboxChange={() =>
+                    setNetworkForm({
+                        ...networkForm,
+                        includeNeighboursWithStructuralData: !networkForm.includeNeighboursWithStructuralData,
+                    })
+                }
+            />
+
+            <Label forText="uploadAnnotations" labelText="Upload your annotations in JSON format" />
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/NetworkForm.tsx:73]: i18n.t

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/SpeciesSelect.tsx [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    "musMusculus",
+    "mycobacteriumTuberculosis",
+    "mycoplasmaPneumoniae",
+    "plasmodiumFalciparum",
+    "rattusNorvegicus",
+    "saccharomycesCerevisiae",
+    "schizosaccharomycesPombe",
+    "synechocystis",
+    "treponemaPallidum",
+] as const;
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/SpeciesSelect.tsx]: That's a collection of names that's not really worth typing it. Note that finally you just send the parent the string (`onSpecieChange: (e: string) => void;`), so we can keep it as an string within the component. I'll make a proposal commit about this.

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/UniProtAccessionTextArea.tsx [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    const { textareaValue, onTextareaChange } = props;
+    return (
+        <textarea
+            id="uniProtAccession"
+            name="uniProtAccession"
+            rows={5}
+            className="form-control"
+            value={textareaValue}
+            onChange={e => onTextareaChange(e.target.value)}
+        ></textarea>
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/UniProtAccessionTextArea.tsx:19]: Small components are on the limit. Do we need a separate component or not? I think it's ok both ways.

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/UniProtAccessionTextArea.tsx [hunk 16]
```diff
@@ -0,0 +1,23 @@
+import React from "react";
+import "./Network.css";
+
+interface UniProtAccessionTextAreaProps {
+    textareaValue: string;
+    onTextareaChange: (e: string) => void;
+}
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/UniProtAccessionTextArea.tsx:7]: As the component refers to a text area, its props can be simply `value` and `onChange`.

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ToolsButton.tsx [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+export const ToolsButton: React.FC<ToolsButtonProps> = () => {
+    const [isNetworkOpen, { enable: openNetwork, disable: closeNetwork }] = useBooleanState(false);
+    const items: DropdownProps["items"] = [{ text: i18n.t("Network"), id: "network" }];
+
+    return (
+        <>
+            <Dropdown
+                text={i18n.t("Tools")}
+                items={items}
+                onClick={() => openNetwork()}
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/protvista/ToolsButton.tsx:18]: openNetwork

File: app/assets/javascripts/3dbio_viewer/src/webapp/components/network/SpeciesSelect.tsx [hunk 18]
```diff
@@ -0,0 +1,71 @@
+import React from "react";
+import i18n from "../../utils/i18n";
+import "./Network.css";
+
+export interface SpeciesSelectProps {
+    value: string;
+    onSpecieChange: (e: string) => void;
```
Comment (tokland) [inline app/assets/javascripts/3dbio_viewer/src/webapp/components/network/SpeciesSelect.tsx]: I think `species` it's also the singular form.

**Review by tokland — CHANGES_REQUESTED** [review]
Looks good! I've made a lot of suggestions and comments, but all of them are minor. Check the proposal commits for the species selector. 

Just one important thing: you have to move the changes to package.json from the other PR, right? dropzone, updated react typings and so on are missing here.

**Review by jvfdunkley — COMMENTED** [review]
All changes look good.

**Review by tokland — APPROVED** [review]
All good!

**Pull request merged**